### ISSUE TITLE
Automatic differentiation with real values

### ISSUE THREAD

**Issue by padreati** [padreati/rapaio#251]
Implement an automatic differentiation engine with real values. This implementation will be the draft revision for automatic differentiation on tensors which later will be used in neural networks. 

This includes the building block for autodiff and some additional tools like loss functions and optimization strategies.

**Comment by padreati** [comment]
In experimental mode, but will be closed since the linear algebra autograd should cover that
